File: PrincipiosDeProgramacionProjects/Perlprojects/redes.pl
#!/usr/bin/perl

use strict;
use warnings;

# Lee las conexiones desde la entrada estándar y crea la matriz de adyacencia
sub crear_matriz_adyacencia {
    my %matriz;

    while (1) {
        print "Ingrese la conexión (o presione Enter para terminar): ";
        my $linea = <STDIN>;
        chomp $linea;

        last if $linea eq ''; # Termina la entrada al presionar Enter

        my ($nodo1, $nodo2) = split /\s+/, $linea;
        
        # Marca la conexión en ambas direcciones (la red es no dirigida)
        $matriz{$nodo1}{$nodo2} = 1;
        $matriz{$nodo2}{$nodo1} = 1;
    }

    return \%matriz;
}

# Calcula e imprime el grado y el coeficiente de clustering para cada nodo
sub calcular_y_mostrar_resultados {
    my ($matriz) = @_;

    for my $nodo (sort keys %$matriz) {
        my $grado = scalar keys %{ $matriz->{$nodo} };
        my $coeficiente = calcular_coeficiente_clustering($matriz, $nodo);

        printf "Nodo: %s, Grado: %d, Coeficiente de Clustering: %.2f\n", $nodo, $grado, $coeficiente;
    }
}

# Calcula el coeficiente de clustering para un nodo específico
sub calcular_coeficiente_clustering {
    my ($matriz, $nodo) = @_;

    my @vecinos = keys %{ $matriz->{$nodo} };
    my $num_conexiones = 0;
    
    for my $vecino1 (@vecinos) {
        for my $vecino2 (@vecinos) {
            next if $vecino1 eq $vecino2;

            # Incrementa el número de conexiones si hay una conexión entre los vecinos
            $num_conexiones++ if $matriz->{$vecino1}{$vecino2};
        }
    }

    my $grado = scalar @vecinos;
    return $grado > 1 ? 2 * $num_conexiones / ($grado * ($grado - 1)) : 0;
}

# Programa principal
sub main {
    my $matriz = crear_matriz_adyacencia();
    calcular_y_mostrar_resultados($matriz);
}

main(); # Llama al programa principal

File: PrincipiosDeProgramacionProjects/Cprojects/redes.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_NODOS 100

// Estructura para representar la red no dirigida
struct RedNoDirigida {
    int matriz[MAX_NODOS][MAX_NODOS];
    int numNodos;
};

// Prototipos de funciones
void inicializarRed(struct RedNoDirigida *red);
void agregarConexion(struct RedNoDirigida *red, int nodo1, int nodo2);
void calcularYMostrarResultados(struct RedNoDirigida *red);
float calcularCoeficienteClustering(struct RedNoDirigida *red, int nodo);

int main() {
    struct RedNoDirigida red;
    inicializarRed(&red);

    printf("Ingrese las conexiones (ingrese -1 para finalizar):\n");

    while (1) {
        int nodo1, nodo2;
        printf("Conexión (nodo1 nodo2): ");
        scanf("%d", &nodo1);

        if (nodo1 == -1) {
            break;
        }

        scanf("%d", &nodo2);
        agregarConexion(&red, nodo1, nodo2);
    }

    calcularYMostrarResultados(&red);

    return 0;
}

// Inicializa la matriz de adyacencia de la red
void inicializarRed(struct RedNoDirigida *red) {
    red->numNodos = 0;

    for (int i = 0; i < MAX_NODOS; i++) {
        for (int j = 0; j < MAX_NODOS; j++) {
            red->matriz[i][j] = 0;
        }
    }
}

// Agrega una conexión entre nodos a la matriz de adyacencia
void agregarConexion(struct RedNoDirigida *red, int nodo1, int nodo2) {
    red->matriz[nodo1][nodo2] = 1;
    red->matriz[nodo2][nodo1] = 1;

    // Actualiza el número total de nodos si es necesario
    red->numNodos = (nodo1 > red->numNodos) ? nodo1 : red->numNodos;
    red->numNodos = (nodo2 > red->numNodos) ? nodo2 : red->numNodos;
}

// Calcula e imprime el grado y el coeficiente de clustering para cada nodo
void calcularYMostrarResultados(struct RedNoDirigida *red) {
    for (int nodo = 1; nodo <= red->numNodos; nodo++) {
        int grado = 0;

        for (int i = 1; i <= red->numNodos; i++) {
            grado += red->matriz[nodo][i];
        }

        float coeficiente = calcularCoeficienteClustering(red, nodo);

        printf("Nodo: %d, Grado: %d, Coeficiente de Clustering: %.2f\n", nodo, grado, coeficiente);
    }
}

// Calcula el coeficiente de clustering para un nodo específico
float calcularCoeficienteClustering(struct RedNoDirigida *red, int nodo) {
    int numVecinos = 0;
    int numConexiones = 0;

    for (int i = 1; i <= red->numNodos; i++) {
        if (red->matriz[nodo][i]) {
            numVecinos++;

            for (int j = i + 1; j <= red->numNodos; j++) {
                if (red->matriz[i][j] && red->matriz[nodo][j]) {
                    numConexiones++;
                }
            }
        }
    }

    float grado = numVecinos;
    return (grado > 1) ? (2.0 * numConexiones) / (grado * (grado - 1)) : 0.0;
}
